packages/amplify_core/lib/src/types/models/mipr/schema_type.dart

```diff
@@ -0,0 +1,644 @@
+  /// {@endtemplate}
+  @BuiltValueEnumConst(wireName: 'AWSJSON')
+  static const AppSyncScalar awsJson = _$awsJson;
+
+  /// {@template amplify_core.models.appsync_scalar_aws_phone}
+  /// A phone number.
+  ///
+  /// This value is stored as a string. Phone numbers can contain either spaces
+  /// or hyphens to separate digit groups. Phone numbers without a country code
+  /// are assumed to be US/North American numbers adhering to the
+  /// [North American Numbering Plan (NANP)](https://en.wikipedia.org/wiki/North_American_Numbering_Plan).
+  /// {@endtemplate}
+  @BuiltValueEnumConst(wireName: 'AWSPhone')
+  static const AppSyncScalar awsPhone = _$awsPhone;
+
+  /// {@template amplify_core.models.appsync_scalar_aws_url}
+  ///
+  /// {@endtemplate}
+  @BuiltValueEnumConst(wireName: 'AWSURL')
+  static const AppSyncScalar awsUrl = _$awsUrl;
+
+  /// {@template amplify_core.models.appsync_scalar_string}
+  /// A URL as defined by [RFC 1738](https://tools.ietf.org/html/rfc1738).
+  ///
+  /// For example, `https://www.amazon.com/dp/B000NZW3KC/` or `mailto:example@example.com`.
+  /// URLs must contain a schema (`http`, `mailto`) and can't contain two
+  /// forward slashes (`//`) in the path part.
+  /// {@endtemplate}
+  @BuiltValueEnumConst(wireName: 'AWSIPAddress')
+  static const AppSyncScalar awsIpAddress = _$awsIpAddress;
+
+  /// The values of [AppSyncScalar].
+  static BuiltSet<AppSyncScalar> get values => _$AppSyncScalarValues;
+
+  /// The [AppSyncScalar] value corresponding to [name].
+  static AppSyncScalar valueOf(String name) => _$AppSyncScalarValueOf(name);
+
+  /// Bridge for legacy codegen.
+  // TODO(dnys1): Remove with platform channel code
+  static AppSyncScalar? fromLegacyType(ModelFieldTypeEnum type) {
+    switch (type) {
+      case ModelFieldTypeEnum.string:
+        return string;
+      case ModelFieldTypeEnum.int:
+        return int_;
+      case ModelFieldTypeEnum.double:
+        return float;
+      case ModelFieldTypeEnum.date:
+        return awsDate;
+      case ModelFieldTypeEnum.dateTime:
+        return awsDateTime;
+      case ModelFieldTypeEnum.time:
+        return awsTime;
+      case ModelFieldTypeEnum.timestamp:
+        return awsTimestamp;
+      case ModelFieldTypeEnum.bool:
+        return boolean;
+      case ModelFieldTypeEnum.enumeration:
+      case ModelFieldTypeEnum.model:
+      case ModelFieldTypeEnum.collection:
+      case ModelFieldTypeEnum.embedded:
+      case ModelFieldTypeEnum.embeddedCollection:
+        return null;
+    }
+  }
+
+  /// The serializer for [AppSyncScalar].
+  static Serializer<AppSyncScalar> get serializer => _$appSyncScalarSerializer;
+}
+
+/// {@template amplify_core.models.schema_type}
+/// An Amplify schema type, either a scalar, model, non-model, or enum.
+///
+/// - [ScalarType]
+/// - [ModelType]
+/// - [NonModelType]
+/// - [EnumType]
+/// - [ListType]
+/// {@endtemplate}
+abstract class SchemaType with AWSSerializable<Map<String, Object?>> {
```

Comment:
Not sure if `SchemaType` is more accurate name than `FiledType`. To me "schema" is much more broad covering object specification, query/mutation selection set etc.

Reply:
FieldType is not fully accurate though. These types will be used in other places beyond a field context.

Reply:
SchemaType = a type defined in the schema

Reply:
Would it make more sense to remove `isRequired` from `SchemaType` and wrap it in a `ModelFieldType` like it was before. I started with this but was trying to consolidate:

Reply:
Actually the problem was that describing recursive types was very difficult. For example, it wasn't possible before to distinguish `[String]!` from `[String!]` or `[String!]!`

Reply:
>Would it make more sense to remove isRequired from SchemaType and wrap it in a ModelFieldType like it was before? I started with this but was trying to consolidate.

Yes I think that would make it clearer, not mixing a property with a `field`, so "SchemaType = a type defined in the schema" would make more sense 👍

Reply:
Okay, I mean I don't think there's anything field-specific here. Maybe the confusion is just that it's called `isRequired`. If it was `nonNull` or something, it might be easier to understand. But I do think this is the best way to represent a type in any context.

Reply:
I can try again with separating the two. I guess that might work since lists are only ever found in a field context.

Reply:
It would basically require wrapping SchemaType like this, removing `isRequired` and `ListType` from `SchemaType`. So `SchemaType` would represent just the name and kind of the type basically.

Thoughts?

```dart
abstract class ModelFieldType {
  const ModelFieldType();

  bool get isRequired;
}

class ModelFieldBaseType extends ModelFieldType {
  const ModelFieldBaseType(this.schemaType, {required this.isRequired});

  final SchemaType schemaType;

  @override
  final bool isRequired;
}

class ModelFieldListType extends ModelFieldType {
  const ModelFieldListType(this.elementType, {required this.isRequired});

  final ModelFieldType elementType;

  @override
  final bool isRequired;
}
```

Reply:
Resolved offline - decided to stay with SchemaType as-is